### PR TITLE
Bug 309: Add integration tests for pom's with different names

### PR DESCRIPTION
**Pull request by DieterMai**
- Add integration test for goal update-pom for poms named pom.xml
- Add integration test for goal update-pom for poms not named pom.xml
- Correct wrong file existens check (basedir -> locatePom)

Dieter Mai

Fixes https://github.com/eclipse-tycho/tycho/issues/309

### CODE REVIEW

**Comment by DieterMai** [comment]
This is a new pull request to resolve the conflicts from PR 1346. PR 1346 can be closed

This PR just adds 2 integration tests. one of them fails. due to Issue 309.

The root issue is that the plugin works with base directories of each plugin instead of the actual pom files. Maybe ProjectMetadata could have field for the pom file name.

Maybe i have a look at this another time for an actual fix. For now only the test.

**Comment by laeubi** [comment]
> Maybe i have a look at this another time for an actual fix. 

That would be great, the code for this is quite local, so it seems a good bug for a beginner with the Tycho codebase

**Comment by DieterMai** [comment]
- Added integration test for modules with different pom names
- Implemented support for custom pom names
Please review

File: tycho-versions-plugin/src/main/java/org/eclipse/tycho/versions/UpdatePomMojo.java [hunk 1]
```diff
@@ -46,7 +47,8 @@ public class UpdatePomMojo extends AbstractMojo {
     public void execute() throws MojoExecutionException, MojoFailureException {
         synchronized (LOCK) {
             try {
-                pomReader.addBasedir(session.getCurrentProject().getBasedir(), true);
+                var project = session.getCurrentProject();
```
Comment (laeubi) [inline tycho-versions-plugin/src/main/java/org/eclipse/tycho/versions/UpdatePomMojo.java:50]: You can inject the project with

```
@Parameter(defaultValue = "${project}", required = true, readonly = true)
private MavenProject project;
```

into the mojo

File: tycho-versions-plugin/src/main/java/org/eclipse/tycho/versions/engine/ProjectMetadataReader.java [hunk 2]
```diff
@@ -48,49 +48,41 @@ public void reset() {
...
-        for (ModelProcessor modelProcessor : modelprocessors) {
-            File locatePom = modelProcessor.locatePom(basedir);
-            if (basedir.exists()) {
-                pomFile = locatePom;
-                break;
-            }
-        }
-        if (pomFile == null || !pomFile.exists() || pomFile.length() == 0) {
-            log.warn("No pom file found at " + basedir);
+        if (isValidPomFile(pomFile)) {
```
Comment (laeubi) [inline tycho-versions-plugin/src/main/java/org/eclipse/tycho/versions/engine/ProjectMetadataReader.java:72]: Should this not be
```suggestion
        if (!isValidPomFile(pomFile)) {
```
?

Comment (DieterMai) [reply tycho-versions-plugin/src/main/java/org/eclipse/tycho/versions/engine/ProjectMetadataReader.java:72]: No, the condition is correct. But the method name is wrong. it should be "isInvalidPomFile(...)"

File: tycho-its/src/test/java/org/eclipse/tycho/test/versionsplugin/TychoVersionsPluginTest.java [hunk 3]
```diff
@@ -79,4 +80,115 @@ public void updateTargetVersionTest() throws Exception {
...
+	 * </p>
+	 * This was created in response to
+	 * <a href="https://github.com/eclipse-tycho/tycho/issues/309">issue 309</a>.
+	 */
+	@Test
+	public void testUpdatePomWithImplicitPomName() throws Exception {
+		String POM_NAME = "pom.xml";
+		String MANIFEST_VERSION = "2.0.0";
+
+		Verifier verifier = getVerifier("tycho-version.update-pom.pomNamedPomXml", false);
```
Comment (laeubi) [inline tycho-its/src/test/java/org/eclipse/tycho/test/versionsplugin/TychoVersionsPluginTest.java:98]: Can you please organize the projects in subfolders like this?

```suggestion
		Verifier verifier = getVerifier("tycho-version-plugin/update-pom/pomNamedPomXml", false);
```

File: tycho-its/src/test/java/org/eclipse/tycho/test/versionsplugin/TychoVersionsPluginTest.java [hunk 4]
```diff
@@ -79,4 +80,115 @@ public void updateTargetVersionTest() throws Exception {
...
+	 * </p>
+	 * This was created in response to
+	 * <a href="https://github.com/eclipse-tycho/tycho/issues/309">issue 309</a>.
+	 */
+	@Test
+	public void testUpdatePomOfPomThatIsNamedPomXml() throws Exception {
+		String POM_NAME = "pom.xml";
+		String MANIFEST_VERSION = "2.0.0";
+
+		Verifier verifier = getVerifier("tycho-version.update-pom.pomNamedPomXml", false);
```
Comment (laeubi) [inline tycho-its/src/test/java/org/eclipse/tycho/test/versionsplugin/TychoVersionsPluginTest.java:120]: same as above...

File: tycho-its/src/test/java/org/eclipse/tycho/test/versionsplugin/TychoVersionsPluginTest.java [hunk 5]
```diff
@@ -79,4 +80,115 @@ public void updateTargetVersionTest() throws Exception {
...
+	 * </p>
+	 * This was created in response to
+	 * <a href="https://github.com/eclipse-tycho/tycho/issues/309">issue 309</a>.
+	 */
+	@Test
+	public void testUpdatePomsOfModularPom() throws Exception {
+		String POM_NAME = "aggregate";
+		String MANIFEST_VERSION = "2.0.0";
+
+		Verifier verifier = getVerifier("tycho-version.update-pom.modularPom", false);
```
Comment (laeubi) [inline tycho-its/src/test/java/org/eclipse/tycho/test/versionsplugin/TychoVersionsPluginTest.java:169]: Similar here
```suggestion
		Verifier verifier = getVerifier("tycho-version-plugin/update-pom/modularPom", false);
```

**Review by laeubi — APPROVED** [review]
This looks fine, can you please squash all commit into one, I think this is then ready to be merged.

**Comment by laeubi** [comment]
@DieterMai thanks for providing tests and fixes!

**Pull request merged**